agent/invocation/state_machine.rb

```diff
@@ -0,0 +1,105 @@
+require "forwardable"
+require "micromachine"
+require_relative "../agent"
+
+module FastlaneCI::Agent
+  ##
+  # StateMachine defines the state and transition events of an Invocation
+  #
+  # This module is implemented to be `prepended` into a class.
+  #
+  # Methods for each transition are defined, and will only be called in the containing class
+  # if the transition is valid.
+  #
+  # On every transition `send_state` will be attempted to be called.
+  module StateMachine
+    include Logging
+    extend Forwardable
+
+    # forward methods to the state machine
+    def_delegators :state_machine, :state, :events, :states
+
+    def state_machine
+      @state_machine ||= MicroMachine.new("pending").tap do |fsm|
+        fsm.when(:run,     "pending"   => "running")
+        fsm.when(:finish,  "running"   => "finishing")
+        fsm.when(:succeed, "finishing" => "succeeded")
+        fsm.when(:reject,  "running"   => "rejected")
+        fsm.when(:fail,    "running"   => "failed")
+        fsm.when(:throw,   "pending"   => "broken",
+                           "running"   => "broken",
+                           "finishing" => "broken")
+
+        # send update whenever we transition states.
+        fsm.on(:any) do |event, payload|
+          send_state(event, payload)
+        end
+      end
+    end
+
+    def send_state(event, payload)
+      logger.debug("Event `#{event}` causing state change to #{state}. #{payload}")
+      super if defined?(super)
```

Comment:
This seems like a code smell calling `super` only if it's defined. What cases would `super` be defined and what cases would it not be defined?

Reply:
For `send_state`, super should always be defined, so I will get rid of this line. Thanks for calling that out.

For the other state transition methods (`run`, `fail`, etc.) , the module prepending `StateMachine` may or may not define any steps to take during the transition. Maybe we should just implement empty methods such as `succeed` and remove that `defined?` check. What do you think?

Reply:
Ahh, okay - that makes sense. I have a mild preference towards the empty methods, but either way sounds good. If we decide on going with the current implementation, could we add a comment for future viewers (and future me because I'll probably forget 😂bahaha)

Reply:
@AndrewMcBurney updated with your suggestion of having empty methods

Reply:
Thank you very much @snatchev! Looks great!